File: Security/hasher.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <pthread.h>
#include <stdbool.h>

struct Hasher {
    pthread_mutex_t lock;
    uint64_t hash;
};

void hasher_init(struct Hasher *hasher) {
    pthread_mutex_init(&hasher->lock, NULL);
    hasher->hash = 0;
}

void hasher_write(struct Hasher *hasher, const char *bytes, size_t len) {
    pthread_mutex_lock(&hasher->lock);
    for (size_t i = 0; i < len; i++) {
        hasher->hash = (hasher->hash * 31) + bytes[i];
    }
    pthread_mutex_unlock(&hasher->lock);
}

uint64_t hasher_finish(struct Hasher *hasher) {
    pthread_mutex_lock(&hasher->lock);
    uint64_t hash = hasher->hash;
    pthread_mutex_unlock(&hasher->lock);
    return hash;
}

uint64_t hash_password(const char *password) {
    struct Hasher hasher;
    hasher_init(&hasher);
    hasher_write(&hasher, password, strlen(password));
    return hasher_finish(&hasher);
}

int main(void) {
    char *users[2][2] = {
        { "secret", "johndoe" },
        { "password", "janedoe" },
    };

    for (size_t i = 0; i < 2; i++) {
        uint64_t hash = hash_password(users[i][0]);
        printf("Utilisateur %s : %lu\n", users[i][1], hash);
    }

    return 0;
}

File: Security/hasher.rs
use std::collections::HashMap;
use std::hash::{Hash, Hasher};
use std::sync::{Arc, Mutex};

struct HasherWrapper(Arc<Mutex<Hasher>>);

impl Hasher for HasherWrapper {
    fn finish(&self) -> u64 {
        self.0.lock().unwrap().finish()
    }

    fn write(&mut self, bytes: &[u8]) {
        self.0.lock().unwrap().write(bytes);
    }
}

fn hash_password(password: &str) -> u64 {
    let hasher = std::collections::hash_map::DefaultHasher::new();
    let hasher = HasherWrapper(Arc::new(Mutex::new(hasher)));
    password.hash(&hasher);
    hasher.finish()
}

fn main() {
    let mut users: HashMap<u64, String> = HashMap::new();

    users.insert(hash_password("secret"), "johndoe".to_string());
    users.insert(hash_password("password"), "janedoe".to_string());

    println!("Utilisateurs : {:?}", users);
}
